linked_list/linked_list.py

```diff
@@ -1,79 +1,118 @@
+        current = self.head
+        while current:
+            count += 1
+            current = current.next
+
+        return count
 
-    # method that returns the value at a given index in the linked list
-    # index count starts at 0
-    # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def get_at_index(self, index):
-        pass
+        current_index = 0
+        current = self.head
+        while current_index < index and current:
+            current_index += 1
+            current = current.next
+
+        if current_index == index and current:
+            return current.value
+        return None
 
-    # method that returns the value of the last node in the linked list
-    # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def get_last(self):
-        pass
+        if not self.head:
+            return None
+
+        current = self.head
+        while current.next:
+            current = current.next
 
-    # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+        return current.value
+
+    # Time Complexity: O(n); more than 1 approach; same complexity 
+    # Space Complexity: O(n); more than 1 approach; same complexity 
     def add_last(self, value):
-        pass
+        if not self.head:
+            self.add_first(value)
+        else:
+            new_node = Node(value)
+            current = self.head
+            while current.next:
+                current = current.next
+            current.next = new_node
+
 
-    # method to return the max value in the linked list
-    # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+        max = self.head.value
+        while current:
+            if current.value > max:
+                max = current.value
+            current = current.next
+
+        return max
 
-    # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def delete(self, value):
-        pass
+        if not self.head:
+            return
+
+        if value == 0:
+            self.head = self.head.next
+
+        current = self.head
+        current_index = 0
+        while current.next and current_index < value - 1:
+            current = current.next
+            current_index += 1
+
+        if current.next:
+            current.next = current.next.next
 
-    # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1) after element is found; O(n) prior

```

Comment:
⏱ Time complexity is O(n) because this function iterates through the entire list and adds them to an array to print out
🪐 Space complexity is O(n) because you are creating an additional array `helper_list` which becomes length n where n is the length of the linked list